File: medium/remove_array_of_nodes_from_linked_list.go
package medium

import (
	easy "github.com/wearypossum4770/dark-coding-challenges/easy"
)

func RemoveArrayOfNodesFromLinkedList(nums []int, head *easy.ImprovedListNode[int]) *easy.ImprovedListNode[int] {
	set := make(map[int]bool, 1e5)
	for _, num := range nums {
		set[num] = true
	}
	dummy := &easy.ImprovedListNode[int]{Val: 0, Next: head}

	for current := dummy; current.Next != nil; {
		if set[current.Next.Val] {
			current.Next = current.Next.Next
		} else {
			current = current.Next
		}
	}
	return dummy.Next
}

File: medium/remove_array_of_nodes_from_linked_list.py
from easy.list_node import BaseListNode


def remove_array_of_nodes_from_linked_list(
    nums: list[int], head: BaseListNode
):
    unique = set(nums)
    dummy = BaseListNode(0, head)
    current = dummy
    while current.next:
        if current.next.val in unique:
            current.next = current.next.next
        else:
            current = current.next
    return dummy.next
